Update display hook to deal with ; in comment

Carreau opened the pull request:
PR #8321 and #8457 combined + tab -> space. 

(will do more fixes if Travis fails)

Commit:
add test file for display hooks

Commit:
Added tests for displayhook

Work in progress. Tests will fail but we will fix them.

Commit:
make the tests fail correctly

Commit:
Update displayhook.py

Updated quiet function in the displayhook.py to properly recognize comment and semicolon by using the tokenize module.

Commit:
Update displayhook.py

Updated to catch IndexError

Carreau committed:
fix indentation by tabs

Carreau committed:
pep8

takluyver commented on IPython/core/displayhook.py:
```diff
@@ -83,9 +84,19 @@ def check_for_underscore(self):
         """Should we silence the display hook because of ';'?"""
         # do not print output if input ends in ';'
+
+        cell = self.shell.history_manager.input_hist_parsed[self.prompt_count]
```
This line needs to be inside the try/except IndexError. The token loop doesn't.

Carreau committed:
Fix test after rebase

 - fix length error and conflict import name
 - invert try/except where necessary
 - use index for token for 2.7 compat
 - don't deep reload some things.

Closes #8321 and closes #8457

Carreau commented:
Rebased and clean. Waiting for tests.

Carreau commented:
Will close #8002 also. Might need manual closing.

takluyver commented:
Tests passing. In we go!

Carreau commented:
Oh, that was quick ! Thanks !
